crates/shim/src/lib.rs

```diff
@@ -84,45 +87,52 @@ pub struct StartOpts {
 }
 
 /// Helper structure that wraps atomic bool to signal shim server when to shutdown the TTRPC server.
+///
 /// Shim implementations are responsible for calling [`Self::signal`].
 #[derive(Clone)]
-pub struct ExitSignal(Arc<AtomicBool>);
+pub struct ExitSignal(Arc<(Mutex<bool>, Condvar)>);
```

Comment:
The issue is on the wait side:
```
    /// Wait for the exit signal to be set.
    fn wait(&self) {
        while !self.0.load(Ordering::Acquire) {
            std::hint::spin_loop();
        }
    }

```

The wait() is effectively a busy loop, so it may waste huge CPU cycles.